### PR TITLE
[Elastic Agent] Add params `prefer_v2_templates=true` on the ES output

### PR DESCRIPTION
**Pull request by ph**
To make sure we use the v2 templates we must send on bulk request the
params `prefer_v2_templates=true` if not it will default to v1 and will
not try to use the v2 template at all.

Reference: #17809

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

File: x-pack/elastic-agent/pkg/agent/application/decorators.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+				},
+			},
+		},
+	}
+
+	programList := make([]program.Program, 0, len(programsToRun))
+
+	for _, program := range programsToRun {
+		if _, found := transpiler.Lookup(program.Config, elasticsearchOutputKey); !found {
+			programList = append(programList, program)
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/agent/application/decorators.go:36]: should it continue to next one?

**Comment by ph** [comment]
@ruflin @michalpristas this commit https://github.com/elastic/beats/pull/18318/commits/0403d3a689936689b4585fd12116e9b7148d6053 fixes the v2 problem. I need to get that approved by beats @urso and make a separate PRs for it.

File: libbeat/esleg/eslegclient/bulkapi.go [hunk 2]
```diff
@@ -225,3 +229,18 @@ func bulkEncode(log *logp.Logger, out BulkWriter, body []interface{}) error {
...
+	}
+
+	if m2 != nil {
+		for k, v := range m2 {
+			mergedParams[k] = v
+		}
+	}
+
+	return mergedParams
+}
```
Comment (michalpristas) [inline libbeat/esleg/eslegclient/bulkapi.go:246]:
```suggestion
func mergeParams(mm ...map[string]string) map[string]string {
	mergedParams := make(map[string]string, 0)
	for _, m := range mm {
		for k, v := range m {
			mergedParams[k] = v
		}
	}
	return mergedParams
}
```

Comment (urso) [inline libbeat/esleg/eslegclient/bulkapi.go:246]: alternative proposal (reduces number of allocs and prepares size of map):

```
if len(m1) == 0 {
  return m2
}
if len(m2) == 0 {
  return m1
}
merged := make(map[string]string, len(m1) + len(m2))
for k, v := range m1 {
  merged[k] = v
}
for k, v := range m2 {
  merged[k] = v
}
```

**Comment by ph** [comment]
@michalpristas I've created https://github.com/elastic/beats/pull/18326 for beats itself, @urso is reviewing it.

File: libbeat/esleg/eslegclient/bulkapi.go [hunk 3]
```diff
@@ -99,6 +101,8 @@ func (conn *Connection) SendMonitoringBulk(
 		return nil, nil
 	}
 
+	mergedParams := mergeParams(conn.ConnectionSettings.Parameters, params)
```
Comment (ruflin) [inline libbeat/esleg/eslegclient/bulkapi.go:104]: Looks like we have the same logic several times, perhaps worth cleaning up in the future ;-)

File: libbeat/esleg/eslegclient/bulkapi.go [hunk 4]
```diff
@@ -71,14 +71,16 @@ func (conn *Connection) Bulk(
 		return 0, nil, nil
 	}
 
+	mergedParams := mergeParams(conn.ConnectionSettings.Parameters, params)
```
Comment (ruflin) [inline libbeat/esleg/eslegclient/bulkapi.go:74]: Nit: I would merge the params more closely before they are used. If there is an error in the bulkEncode we never would have to merge them. Same below.

File: libbeat/esleg/eslegclient/bulkapi.go [hunk 5]
```diff
@@ -225,3 +229,23 @@ func bulkEncode(log *logp.Logger, out BulkWriter, body []interface{}) error {
 	}
 	return nil
 }
+
+func mergeParams(m1, m2 map[string]string) map[string]string {
```
Comment (ruflin) [inline libbeat/esleg/eslegclient/bulkapi.go:233]: Would be nice to have some unit tests for `mergeParams`.

Comment (ph) [reply libbeat/esleg/eslegclient/bulkapi.go:237]: I will do it as part of the comment in the orignal PR.

**Comment by ruflin** [comment]
Tested locally with 7.8 Stack. Works.

**Comment by ph** [comment]
merged

File: libbeat/esleg/eslegclient/bulkapi.go (hunk 3 above)
Comment (ph) [reply libbeat/esleg/eslegclient/bulkapi.go:108]: I wanted to make it close to the HTTP, but this is not currently possible with the way the HTTP client is composed today we lose context where the we do the actual URL generation. :/

**Pull request merged**